### ISSUE TITLE
PubSub AckDeadline of 10-14 seconds causes subscriber failure

### ISSUE THREAD

**Issue by psyonix-ecarpenter** [googleapis/google-cloud-dotnet#5866]
#### Issue Summary

We wanted to be able to control our `SubscriberClient.Settings` values without needing an image rebuild/deploy, so made changes in code to expose these as environment variables via the deployment's YAML. Previously this was done manually and required an image rebuild/deploy. The two notable fields we're attempting to control are `AckDeadline` and `FlowControlSettings` -> `maxOutstandingElementCount`. Once the relevant image with changes is deployed our pods come up and recognize the env vars (tested by printing ingested env var values at startup) however there is no message throughput, defined here as no messages being pulled from the PubSub server. This was immediately rectified by rolling back the deployment to the prior image version.

Different values for these two fields caused no change in behavior once deployed. By _not_ specifying any fields except for `UseLegacyFlowControl = true` pods come up and happily process messages. This of course isn't great because we don't have dynamic/granular control of our client's settings.

#### Environment details

- Kubernetes Deployment
- .NET version: Core 2.1
- Package name and version: Google.Cloud.PubSub.V1 2.2.0
- Node details
  - Kernel version: 4.19.112+ 
  - OS image: Container-Optimized OS from Google
  - Container runtime version: docker://19.3.1
  - kubelet version: v1.15.12-gke.20

#### Steps to reproduce

  1. Set env vars in deployment YAML
 
```
env:
- name: MAXMESSAGES
  value: 1000 
- name: ACKDEADLINESECONDS
  value: 10
```
  2. (crudely) Pull env vars in

```
public static long? MaxOutstandingElementCount { get; set; }
public static System.TimeSpan? AckDeadlineSeconds { get; set; }

//max messages
if (System.Environment.GetEnvironmentVariable("MAXMESSAGES") != null)
{
	MaxOutstandingElementCount = long.Parse(System.Environment.GetEnvironmentVariable("MAXMESSAGES"));
} else
{
	MaxOutstandingElementCount = null;
}

//ack deadline seconds
if (System.Environment.GetEnvironmentVariable("ACKDEADLINESECONDS") != null)
{
	AckDeadlineSeconds = System.TimeSpan.FromSeconds(int.Parse(System.Environment.GetEnvironmentVariable("ACKDEADLINESECONDS")));
} else
{
	AckDeadlineSeconds = null;
}
```
  3. Create an async SubscriberClient with `AckDeadline` and `maxOutstandingElementCount` specified:

```
_sub = await SubscriberClient.CreateAsync(SubName, settings: new SubscriberClient.Settings()
{
	AckDeadline = Constants.AckDeadlineSeconds,
	FlowControlSettings = new Google.Api.Gax.FlowControlSettings(
		maxOutstandingElementCount: Constants.MaxOutstandingElementCount,
		),
});

await _sub.StartAsync(async (msg, cancellationToken) =>
{
	//message processing happens here
});
```

4. Create docker image
```
FROM gcr.io/{snipped}/dotnet-build-base:2.1 AS build
WORKDIR /build/ThisApp

COPY ThisApp.csproj /build/ThisApp/

COPY . /build/ThisApp

RUN dotnet clean -c Release -f netcoreapp2.1 && \
    dotnet publish -c Release -f netcoreapp2.1 -o /build/dist

FROM microsoft/dotnet:2.1-runtime
WORKDIR /app

COPY --from=build /build/dist /app

USER nobody
ENTRYPOINT [ "dotnet", "ThisApp.dll" ]
```
5. Set deployment image
```
kubectl set image deployment/ThisDeployment app={image}
````

**Comment by jskeet** [comment]
Please could you add logging to record what values for the environment-configured variables are actually getting?

There's nothing in the client library that should care where the values in the settings came from. If some combination of values is causing a problem, it should be possible (and simpler) to demonstrate the problem by hard-coding them in the settings - and I'd expect to be able to see that problem without deploying to k8s, but by just running locally.

Are you able to reproduce the problem in a simple console app with hard-coded settings? That will make it much simpler for me to help you.

**Comment by psyonix-ecarpenter** [comment]
Here's what we emit via Console.WriteLine and see in logs when a pod comes up:
```
INFO - {app} - up and running! Subscription = {sub}, AckDeadline = 10, MaxMessages = 1000
```

An interesting point about hardcoding - in prior images with hardcoded settings we didn't experience this issue, only when we moved to pulling them in from env. I'll work on getting this to repro locally and post back here with my findings.

**Comment by jskeet** [comment]
If you *do* find that hard-coding works, then I suspect you've got a timing issue - where you're starting the subscriber before the static fields have been populated. (We can't currently see that code.)

Reproducing locally will be very helpful indeed - thanks. (We'll get there :)

**Comment by psyonix-ecarpenter** [comment]
Running our consumer locally I was able to consistently reproduce this issue and got an error back as well. Here's some pseudo'ish code of main:

```
//added for testing env vars
Environment.SetEnvironmentVariable("MAXMESSAGES", "1000");
Environment.SetEnvironmentVariable("ACKDEADLINESECONDS", "10");

//normal main
Constants.Initialize();  //<-- env vars are pulled in
var processor = new Processor(bq);
processor.Execute().Wait();
```

I specified settings (initialized within `Processor()`) as:
```
_sub = await SubscriberClient.CreateAsync(SubName, settings: new SubscriberClient.Settings()
{
	AckDeadline = Constants.AckDeadlineSeconds,
	FlowControlSettings = new Google.Api.Gax.FlowControlSettings(
		maxOutstandingElementCount: Constants.MaxOutstandingElementCount,
		maxOutstandingByteCount: Constants.MaxOutstandingByteCount
		),
});
```
And then to determine whether there were messages being processed added a little one-liner into our message processing:
```
Console.WriteLine($"processed MessageID {msg.MessageId}");
```

Interestingly I didn't even get to the point of processing messages before this System.AggregateException exception was thrown:
```
System.AggregateException
  HResult=0x80131500
  Message=One or more errors occurred. (The value needs to translate in milliseconds to -1 (signifying an infinite timeout), 0 or a positive integer less than or equal to Int32.MaxValue.
Parameter name: delay)
  Source=System.Private.CoreLib
  StackTrace:
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at App.Program.Main(String[] args) in C:\repo\App\Program.cs:line 26

Inner Exception 1:
ArgumentOutOfRangeException: The value needs to translate in milliseconds to -1 (signifying an infinite timeout), 0 or a positive integer less than or equal to Int32.MaxValue.
Parameter name: delay
```

Commenting out `AckDeadline` in settings allows the app to continue/process messages successfully.

Pulling in the related env var `ACKDEADLINESECONDS` is done as:
```
AckDeadlineSeconds = System.TimeSpan.FromSeconds(int.Parse(System.Environment.GetEnvironmentVariable("ACKDEADLINESECONDS")));
```

As far as I know, if that env var is set as above to 10 then this should represent a 10 second ack deadline, but based on the stack it appear it may not be? Even more confounding is there does appear to be a working threshold of 15s or greater where the consumer does successfully fire up and process:
- 0-9 -> `(Value 00:00:0{x} should be in range [00:00:10, 00:10:00] Parameter name: AckDeadline)`
- 10-14 -> above exception
- 15 +-> successful program execution

Any thoughts on this? Short-term I'm fine using 15s+ for ack deadlines, but I'm a little puzzled why 10-14s isn't supported even though specifying a < 10s deadline explicitly calls out [00:00:10, 00:10:00] as being permissible.

**Comment by jskeet** [comment]
I will have a look tomorrow. It would still be easier to help with a complete example instead of snippets though.

**Comment by jskeet** [comment]
Oh, and do you have a complete stack trace for that inner exception? That would be very helpful.

**Comment by psyonix-ecarpenter** [comment]
Sure, here's the stack trace:
```
at System.Threading.Tasks.Task.Delay(TimeSpan delay, CancellationToken cancellationToken)  
at Google.Api.Gax.SystemScheduler.Delay(TimeSpan timeSpan, CancellationToken cancellationToken) in /_/Google.Api.Gax/SystemScheduler.cs:line 27   
at Google.Cloud.PubSub.V1.SubscriberClientImpl.SingleChannel.HandleExtendLease(HashSet`1 msgIds, LeaseCancellation cancellation) in /_/apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs:line 1251   
at Google.Cloud.PubSub.V1.SubscriberClientImpl.SingleChannel.HandlePullMessageData(Task`1 moveNextTask) in /_/apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs:line 1099  
at Google.Cloud.PubSub.V1.SubscriberClientImpl.SingleChannel.StartAsync() in /_/apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs:line 923  
at Google.Cloud.PubSub.V1.Tasks.Extensions.<>c__DisplayClass4_0.<<ConfigureAwaitHideErrors>g__Inner|0>d.MoveNext() in /_/apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/Tasks/TaskHelper.cs:line 156
--- End of stack trace from previous location where exception was thrown ---   
at App.Flow.Processor.Execute() in C:\\repo\\App\\Flow\\Processor.cs:line 52
```

**Comment by jskeet** [comment]
Notes to self: the `Delay` exception is from this line:

```csharp
Add(_scheduler.Delay(_autoExtendInterval, _softStopCts.Token), Next(false, () => HandleExtendLease(msgIds, cancellation)));
```

So presumably `_autoExtendInterval` is invalid, although it's not clear what it would be yet.

**Comment by jskeet** [comment]
The auto-extend interval is determined here:

```csharp
_autoExtendInterval = TimeSpan.FromSeconds(_modifyDeadlineSeconds) - (settings.AckExtensionWindow ?? DefaultAckExtensionWindow);
```

Here `_modifyDeadlineSeconds` will be the value you've used for `AckDeadlineSeconds` and the default ack extension window is 15 seconds, so we'd end up with a negative `_autoExtendInterval` - hence the exception.

I suspect that arithmetic is wrong, but this is an area I'm not familiar with.

**Comment by jskeet** [comment]
Okay, I've got a minimal, complete example that demonstrates the problem:

```csharp
using Google.Cloud.PubSub.V1;
using System;
using System.Threading;
using System.Threading.Tasks;

namespace Issue5866
{
    class Program
    {
        static async Task Main(string[] args)
        {
            if (args.Length != 2)
            {
                Console.WriteLine($"Please specify project ID and subscription ID as command-line arguments");
            }
            var settings = new SubscriberClient.Settings
            {
                // Value >= 10s, but < 15s.
                AckDeadline = TimeSpan.FromSeconds(13)
            };
            // These don't need to be real values.
            var subscriptionName = SubscriptionName.FromProjectSubscription(args[0], args[1]);

            var client = await SubscriberClient.CreateAsync(subscriptionName, settings: settings);
            Log("Created client");
            Task clientTask = client.StartAsync(HandleMessage);
            Log("Started client; waiting for 30 seconds");
            await Task.Delay(TimeSpan.FromSeconds(30));
            Log("Stopping client");
            var hardStopToken = new CancellationTokenSource(2000).Token;
            await client.StopAsync(hardStopToken);
            await clientTask;
            Log("Completed");
        }

        static Task<SubscriberClient.Reply> HandleMessage(PubsubMessage message, CancellationToken token)
        {
            Log($"Received message {message.MessageId}");
            return Task.FromResult(SubscriberClient.Reply.Ack);
        }

        static void Log(string text) => Console.WriteLine($"{DateTime.UtcNow:HH:mm:ss}: {text}");
    }
}
```

Steps:

- Create a pubsub topic, and include the default subscription
- Run the above code, passing in the project ID and subscription ID
- Wait for just over 30 seconds
- Observe the exception from HandleExtendLease

**Comment by chrisdunelm** [comment]
This looks like a settings validation (and possibly documentation) issue, rather than a bug in the deadline calculations.

`AckDeadline` is the lease time for pulled messages. `AckExtensionWindow` is the duration before `AckDeadline` when a lease-extension is performed.
Currently it is not being validated that `AckDeadline` is greater than `AckExtensionWindow`.
Because `DefaultAckExtensionWindow` is 15 seconds, this will result in the error being seen if `AckDeadline` < 15 seconds; as is happening here.

I'll create a PR to add this validation.

**Comment by jskeet** [comment]
Until the new version is released (which will be on Monday) I would recommend setting AckDeadline to 20s. This means you'll take out a lease for 20s, but after 5s the lease will be renewed. (If you set AckDeadline to something smaller, you'll currently be renewing the lease almost *immediately*, which I suspect will lead to more lease renewal.)

**Comment by psyonix-ecarpenter** [comment]
That was my thinking as well. Thank you guys for the quick attention and turnaround!

**Comment by jskeet** [comment]
The fix has now been released in 2.3.0. Please let us know if you run into any issues.
